### PR TITLE
Add missing spaces in gdb-remote packets

### PR DESCRIPTION
**Pull request by fjricci**

### CODE REVIEW

**Comment by fjricci** [comment]
Looks like this breaks TestInlineStepping without http://reviews.llvm.org/D20509

**Comment by sas** [comment]
The interesting part is that it works for x86 runs.

On Sun, May 22, 2016 at 12:39 PM, Francis Ricci <notifications@github.com>
wrote:

> Looks like this breaks TestInlineStepping without
> http://reviews.llvm.org/D20509
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/ds2/pull/349#issuecomment-220851618>
>



-- 
Stephane Sezer

**Comment by fjricci** [comment]
So the other interesting part is that qfThreadInfo is only used (by lldb) as a fallback packet when qProcessInfo fails and then qC also fails.

**Comment by fjricci** [comment]
Looks like it's also used as a fallback when it doesn't have a stop reply packet with threads in it. Looks like that's the case we're hitting here.

File: Headers/DebugServer2/GDBRemote/SessionBase.h [hunk 1]
```diff
@@ -30,13 +30,14 @@ class SessionBase : public ProtocolHandler {
...
+  CompatibilityMode _compatMode;
 
 public:
-  SessionBase();
+  SessionBase(CompatibilityMode mode);
   virtual ~SessionBase();
 
 public:
-  virtual CompatibilityMode mode() const = 0;
+  CompatibilityMode mode() const {return _compatMode; };
```
Comment (sas) [inline Headers/DebugServer2/GDBRemote/SessionBase.h]: Hon. Style.

Comment (fjricci) [inline Headers/DebugServer2/GDBRemote/SessionBase.h]: right.

File: Sources/GDBRemote/Session.cpp [hunk 2]
```diff
@@ -34,7 +34,7 @@ namespace ds2 {
 namespace GDBRemote {
 
 Session::Session(CompatibilityMode mode)
-    : _compatMode(mode), _threadsInStopReply(false) {
+    : SessionBase(mode), _threadsInStopReply(false) {
```
Comment (sas) [inline Sources/GDBRemote/Session.cpp]: I'm surprised we weren't calling the `SessionBase()` constructor explicitly before. This is good.

File: Sources/GDBRemote/SessionBase.cpp [hunk 3]
```diff
@@ -139,6 +139,12 @@ bool SessionBase::sendACK() { return _channel->send("+", 1) == 1; }
 
 bool SessionBase::sendNAK() { return _channel->send("-", 1) == 1; }
 
+void SessionBase::addPacketSeparator(std::ostringstream &ss) {
```
Comment (sas) [inline Sources/GDBRemote/SessionBase.cpp]: You should have a comment in this function to explain what we're trying to achieve.

File: Sources/GDBRemote/SessionBase.cpp [hunk 4]
```diff
@@ -158,12 +158,12 @@ bool SessionBase::sendError(ErrorCode code) {
...
-  char message[4];
-  message[0] = 'E';
-  message[1] = NibbleToHex(code >> 4);
-  message[2] = NibbleToHex(code & 15);
-  message[3] = '\0';
-  return send(message);
+  std::ostringstream ss;
+  ss << "E";
+  addPacketSeparator(ss);
+  ss << std::hex << code;
```
Comment (sas) [inline Sources/GDBRemote/SessionBase.cpp]: This is not good, you're changing the packet format. `std::hex` doesn't add `0` at the beginning to always have a two character-long error code.

**Comment by sas** [comment]
@facebook-github-bot shipit

File: Sources/GDBRemote/SessionBase.cpp [hunk 5]
```diff
@@ -139,6 +139,25 @@ bool SessionBase::sendACK() { return _channel->send("+", 1) == 1; }
...
+  switch (_compatMode) {
+  case kCompatibilityModeGDB:
+  case kCompatibilityModeGDBMultiprocess:
+    return " ";
+
+  case kCompatibilityModeLLDB:
+    return "";
+
+  case kCompatibilityModeLLDBThread:
+    DS2BUG("LLDBThreads is an invalid compatibility mode for SessionBase.");
```
Comment (sas) [inline Sources/GDBRemote/SessionBase.cpp]: No ending period !@#$ 😆

**Pull request merged**